crates/services/upgradable-executor/src/instance.rs

```diff
@@ -0,0 +1,592 @@
+    },
+};
+use fuel_core_storage::{
+    column::Column,
+    kv_store::{
+        KeyValueInspect,
+        Value,
+    },
+};
+use fuel_core_types::{
+    blockchain::primitives::DaBlockHeight,
+    fuel_tx::Transaction,
+    fuel_vm::checked_transaction::Checked,
+    services::executor::{
+        Error as ExecutorError,
+        Result as ExecutorResult,
+    },
+};
+use fuel_core_wasm_executor::utils::{
+    pack_exists_size_result,
+    unpack_ptr_and_len,
+    ReturnType,
+};
+use std::{
+    collections::HashMap,
+    sync::Arc,
+};
+use wasmtime::{
+    AsContextMut,
+    Caller,
+    Engine,
+    Func,
+    Linker,
+    Memory,
+    Module,
+    Store,
+};
+
+trait CallerHelper {
+    /// Writes the encoded data to the memory at the provided pointer.
+    fn write(&mut self, ptr: u32, encoded: &[u8]) -> anyhow::Result<()>;
+}
+
+impl<'a> CallerHelper for Caller<'a, ExecutionState> {
+    fn write(&mut self, ptr: u32, encoded: &[u8]) -> anyhow::Result<()> {
+        let memory = self.data_mut().memory.expect("Memory is initialized; qed");
+        let mut store = self.as_context_mut();
+        memory
+            .write(&mut store, ptr as usize, encoded)
+            .map_err(|e| anyhow::anyhow!("Failed to write to the memory: {}", e))
+    }
+}
+
+/// The state used by the host functions provided for the WASM executor.
+struct ExecutionState {
+    /// The memory used by the WASM module.
+    memory: Option<Memory>,
+    next_transaction: HashMap<u32, Vec<Vec<u8>>>,
+    relayer_events: HashMap<DaBlockHeight, Value>,
+}
+
+/// The WASM instance has several stages of initialization.
+/// When the instance is fully initialized, it is possible to run it.
+///
+/// Each stage has a state. Advancing the stage may add a new state that
+/// will later be used by the next stages or a `run` function.
+///
+/// Currently, the order of the definition of the host functions doesn't matter,
+/// but later, the data from previous stages may be used to initialize new stages.
+pub struct Instance<Stage = ()> {
+    store: Store<ExecutionState>,
+    stage: Stage,
+}
+
+/// The stage indicates that the instance is fresh and newly created.
+pub struct Created {
+    linker: Linker<ExecutionState>,
+}
+
+impl Instance {
+    pub fn new(engine: &Engine) -> Instance<Created> {
+        Instance {
+            store: Store::new(
+                engine,
+                ExecutionState {
+                    memory: None,
+                    next_transaction: Default::default(),
+                    relayer_events: Default::default(),
+                },
+            ),
+            stage: Created {
+                linker: Linker::new(engine),
+            },
+        }
+    }
+}
+
+/// This stage adds host functions to get transactions from the transaction source.
+pub struct Source {
+    linker: Linker<ExecutionState>,
+}
+
+impl Instance<Created> {
+    /// Adds host functions for the `source`.
+    pub fn add_source<TxSource>(
```

Comment:
These 3 `add` methods are all really long and have a fair bit of `wasmtime` DSL. Could you break them down to be a series of declarative helper functions? e.g. instead of having the entire `Func::wrap` closure, just have a method like:
```
let peek_next_txs_size_func = self.get_peek_next_txs_size_func(...)
```
`add_source` could be like 7 lines + the return and it would be easier to read.

Reply:
Well, it might make more sense to have the function `get` and the linker call be in single action; the functions doesn't mean much if you don't see how it's being passed to the linker. So might be fewer than 7 lines. LMK if that doesn't make sense.